File: android/app/src/main/java/com/tdrevans/nexttrain/CommuteSchedule.java
package com.tdrevans.nexttrain;

import android.content.Context;
import org.json.JSONArray;
import org.json.JSONObject;

/** Shared journey fetch + leave-by computation for widget and leave reminders. */
public final class CommuteSchedule {

  public static final long STALE_THRESHOLD_MS = 120L * 60L * 1000L;
  public static final long STALE_FETCH_VISIBLE_MS = 45L * 1000L;
  public static final long UPDATING_TIMEOUT_MS = 90L * 1000L;
  public static final long OPPORTUNISTIC_REFRESH_AGE_MS = 12L * 60L * 1000L;
  public static final long OPPORTUNISTIC_REFRESH_DEBOUNCE_MS = 5L * 60L * 1000L;
  static final int PRE_DEPARTURE_PREFETCH_MINUTES = 3;
  public static final String DEGRADED_SECONDARY = CommuteSchedulePreview.DEGRADED_SECONDARY;

  /** Widget/reminder load bag. Extends {@link CommuteScheduleResult} for existing call sites. */
  public static final class Result extends CommuteScheduleResult {}

  private CommuteSchedule() {}

  public static Result load(Context context, boolean allowStaleFallback) {
    Result result = new Result();
    result.refreshedAtMs = WidgetSettingsStore.readLastRefreshMs(context);

    try {
      String settingsJson = WidgetSettingsStore.readSettings(context);
      if (settingsJson == null || settingsJson.isEmpty()) {
        result.empty = true;
        return result;
      }

      result.settings = new JSONObject(settingsJson);

      Result pinnedResult = WidgetPinResolver.loadBestPinnedResult(context, result.settings);
      if (pinnedResult != null) {
        pinnedResult.refreshedAtMs = System.currentTimeMillis();
        WidgetSettingsStore.saveLastRefreshMs(context, pinnedResult.refreshedAtMs);
        pinnedResult.stale = false;
        return pinnedResult;
      }

      result.journey = JourneySelector.selectJourney(result.settings);
      if (result.journey == null) {
        result.journey = JourneySelector.selectActiveRoute(result.settings);
      }
      if (result.journey == null) {
        if (JourneySelector.hasConfiguredJourneys(result.settings)) {
          result.nearbyFallback = true;
          result.empty = false;
        } else {
          result.empty = true;
        }
        return result;
      }

      boolean routeJourney = JourneySelector.isRouteJourney(result.journey);
      result.journeyId = result.journey.optString("id");
      result.route = WidgetDataService.formatRoute(result.journey);
      result.departMode = routeJourney || !result.journey.optBoolean("useLeaveBefore", true);

      int leaveBefore = result.departMode ? 0 : result.journey.optInt("leaveBeforeMinutes", 10);
      result.payload = NextTrainApiClient.fetchNextTrain(
        result.journey.optString("station"),
        result.journey.optString("direction"),
        leaveBefore
      );
      result.refreshedAtMs = System.currentTimeMillis();
      WidgetSettingsStore.saveLastRefreshMs(context, result.refreshedAtMs);
      result.stale = false;
      // Widget face: pin or target only — never true-next (FB-20 / FB-43).
      result.next = JourneyPinHelper.resolvePinnedTrip(result.payload, result.journey);
      fillTripFields(result);
      return result;
    } catch (Exception error) {
      if (allowStaleFallback) {
        long age = System.currentTimeMillis() - result.refreshedAtMs;
        result.stale = result.refreshedAtMs > 0 && age > STALE_THRESHOLD_MS;

        JSONObject cached = WidgetSettingsStore.readSnapshot(context);
        if (cached != null) {
          try {
            result.journeyId = cached.optString("journeyId");
            result.route = cached.optString("route");
            result.empty = cached.optBoolean("empty", false);
          } catch (Exception ignored) {
            result.empty = true;
          }
        } else {
          result.empty = true;
        }
      } else {
        result.empty = true;
      }
      return result;
    }
  }

  static Result tryLoadNearbyPin(Context context, JSONObject settings) throws Exception {
    JSONObject pin = settings.optJSONObject("nearbyPin");
    if (!NearbyPinHelper.isHolding(pin)) {
      return null;
    }

    int leaveBefore = settings.optInt("nearbyLeaveBeforeMinutes", 10);
    PreferredTrainReminder.Target target = NearbyPinHelper.computeTarget(pin, leaveBefore, false);
    if (target == null) {
      return null;
    }

    JSONObject payload = NextTrainApiClient.fetchNextTrain(
      pin.optString("station", ""),
      pin.optString("direction", ""),
      leaveBefore
    );

    Result result = new Result();
    result.settings = settings;
    result.journeyId = NearbyPinHelper.JOURNEY_ID;
    result.route = target.route;
    result.departMode = false;
    result.payload = payload;
    result.refreshedAtMs = System.currentTimeMillis();
    WidgetSettingsStore.saveLastRefreshMs(context, result.refreshedAtMs);
    result.stale = false;

    JSONObject trip = NearbyPinHelper.findTripByDeparture(payload, target.departureIso);
    if (trip == null) {
      trip = NearbyPinHelper.buildSyntheticTrip(pin, leaveBefore);
    }
    result.next = trip;
    fillTripFields(result);
    return result;
  }

  static void fillTripFields(Result result) {
    if (result.next == null) {
      return;
    }
    result.leaveByIso = result.next.optString("leaveBy", "");
    result.departureIso = result.next.optString("departure", "");
    if (result.departureIso.isEmpty()) {
      result.departureIso = result.next.optString("arrival", "");
    }
    result.displayTime = result.next.optString("displayTime", "—");
    result.status = result.next.optString("status", "On Time");
    result.leavePhase = result.next.optString("leavePhase", "calm");
    result.minutesUntilLeave = result.next.optInt("minutesUntilLeave", 0);
  }

  public static String departureKey(Result result) {
    if (result == null || result.journeyId == null || result.departureIso == null) {
      return "";
    }
    if (result.departureIso.isEmpty()) {
      return "";
    }
    return result.journeyId + ":" + result.departureIso;
  }

  public static long parseLeaveByMs(Result result) {
    if (result == null) {
      return 0L;
    }
    return PerthTime.epochMillisFromIso(result.leaveByIso);
  }

  public static JSONObject toWidgetSnapshot(Result result) throws Exception {
    return CommuteScheduleSnapshot.toWidgetSnapshot(result);
  }

  public static JSONObject nearbyFallbackIfOutsideHours(Context context) {
    return CommuteSchedulePreview.nearbyFallbackIfOutsideHours(context);
  }

  public static JSONObject outsideHoursSnapshot(JSONObject settings) throws Exception {
    return CommuteSchedulePreview.outsideHoursSnapshot(settings);
  }

  static JSONObject outsideHoursSnapshot(
    JSONObject settings,
    int nowMinutes,
    int dayOfWeekIso
  ) throws Exception {
    return CommuteSchedulePreview.outsideHoursSnapshot(settings, nowMinutes, dayOfWeekIso);
  }

  public static boolean hasWidgetAccess(JSONObject settings) {
    return true;
  }

  public static boolean hasWidgetAccessFromContext(Context context) {
    try {
      String json = WidgetSettingsStore.readSettings(context);
      if (json == null || json.isEmpty()) {
        return true;
      }
      return hasWidgetAccess(new JSONObject(json));
    } catch (Exception error) {
      return true;
    }
  }

  public static JSONObject widgetLockedSnapshot() throws Exception {
    return CommuteSchedulePreview.widgetLockedSnapshot();
  }

  public static JSONObject repaintSnapshot(JSONObject cached) throws Exception {
    return CommuteScheduleSnapshot.repaintSnapshot(cached);
  }

  /**
   * Soonest not-yet-departed train (classic next). Hero/widget may show a pin instead;
   * use {@link JourneyPinHelper#resolvePinnedTrip} for the active commute face.
   */
  static JSONObject resolveTrueNextTrip(JSONObject payload) {
    if (payload == null) {
      return null;
    }

    JSONArray upcoming = collectUpcomingTrips(payload);
    if (upcoming.length() > 0) {
      for (int index = 0; index < upcoming.length(); index += 1) {
        JSONObject trip = upcoming.optJSONObject(index);
        if (trip == null || hasDepartureMinutePassed(trip)) {
          continue;
        }
        return trip;
      }
      return null;
    }

    JSONObject next = payload.optJSONObject("next");
    if (next != null && !hasDepartureMinutePassed(next)) {
      return next;
    }

    JSONObject following = payload.optJSONObject("following");
    if (following != null && !hasDepartureMinutePassed(following)) {
      return following;
    }

    return null;
  }

  /** Leave By / leave twin only when no preferred, or this trip is at/after preferred. */
  static boolean leaveByArmedForTrip(JSONObject trip, JSONObject journey) {
    if (journey == null || JourneySelector.isRouteJourney(journey)) {
      return false;
    }
    if (!journey.optBoolean("useLeaveBefore", true)) {
      return false;
    }
    if (trip == null) {
      return false;
    }
    if (JourneyPinHelper.isOverrideActiveToday(journey)) {
      return true;
    }
    int preferredMinutes = preferredMinutesForLiveGlance(journey);
    if (preferredMinutes < 0) {
      return true;
    }
    return tripMatchesPreferredOrLater(trip, preferredMinutes, liveHorizonMinutes(journey));
  }

  /** True when the widget live face is an actively pinned train (nearby, override, or commute pin). */
  static boolean isWidgetPinnedFace(JSONObject journey) {
    if (journey == null) {
      return false;
    }
    if (NearbyPinHelper.JOURNEY_ID.equals(journey.optString("id", ""))) {
      return true;
    }
    if (JourneyPinHelper.isOverrideActiveToday(journey)) {
      return true;
    }
    if (JourneySelector.isRouteJourney(journey)) {
      return false;
    }
    return JourneyPinHelper.isJourneyPinnedToday(journey);
  }

  /** True when the widget label should read Pinned (nearby or day override — not preferred target). */
  static boolean isWidgetPinnedChromeLabel(JSONObject journey) {
    if (journey == null) {
      return false;
    }
    if (NearbyPinHelper.JOURNEY_ID.equals(journey.optString("id", ""))) {
      return true;
    }
    return JourneyPinHelper.isOverrideActiveToday(journey);
  }

  static boolean isWidgetLiveFaceLabel(String label) {
    if (label == null || label.isEmpty()) {
      return false;
    }
    String normalized = label.trim().toLowerCase(java.util.Locale.US);
    return normalized.equals("pinned")
      || normalized.equals("target")
      || normalized.equals("pinned train")
      || normalized.equals("target train");
  }

  static String normalizeWidgetLiveFaceLabel(String label) {
    if (label == null || label.isEmpty()) {
      return "Target train";
    }
    String normalized = label.trim().toLowerCase(java.util.Locale.US);
    if (normalized.equals("pinned") || normalized.equals("pinned train")) {
      return "Pinned train";
    }
    if (normalized.equals("target") || normalized.equals("target train")) {
      return "Target train";
    }
    return label;
  }

  /**
   * Live face label: Pinned train (nearby / day override) or Target train (preferred target + default).
   * (Idle outside-hours uses {@link NextCommutePreview#idleWidgetLabel}.)
   */
  static String liveWidgetLabel(JSONObject journey, JSONObject trip) {
    if (trip == null) {
      return "";
    }
    if (isWidgetPinnedChromeLabel(journey)) {
      return "Pinned train";
    }
    return "Target train";
  }

  static String liveWidgetLabelFromSnapshot(JSONObject snapshot) {
    if (snapshot == null) {
      return "Target train";
    }
    if (snapshot.optBoolean("widgetPinnedChrome", false)) {
      return "Pinned train";
    }
    return "Target train";
  }

  /** Preserve Pinned / Target across Updating… / stale paints. */
  static String preservedLiveLabel(JSONObject cached) {
    if (cached == null) {
      return "";
    }
    String label = cached.optString("label", "");
    if (isWidgetLiveFaceLabel(label)) {
      String normalized = normalizeWidgetLiveFaceLabel(label);
      // Legacy snapshots stored preferred-target face as Pinned before widgetPinnedChrome existed.
      if (
        "Pinned train".equals(normalized) &&
        !cached.optBoolean("widgetPinnedChrome", false) &&
        !NearbyPinHelper.JOURNEY_ID.equals(cached.optString("journeyId", ""))
      ) {
        return "Target train";
      }
      return normalized;
    }
    if (!label.isEmpty()) {
      return label;
    }
    return liveWidgetLabelFromSnapshot(cached);
  }

  /** Medium widget orientation when Leave By is hidden for an earlier train. */
  static String preferredHintForJourney(JSONObject journey) {
    int preferredMinutes = preferredMinutesForLiveGlance(journey);
    if (preferredMinutes < 0) {
      return "";
    }
    return "Target " + NextCommutePreview.formatClock(preferredMinutes);
  }

  /** Explicit preferred only — do not fall back to Active from (unlike reminders legacy). */
  static int preferredMinutesForLiveGlance(JSONObject journey) {
    if (journey == null) {
      return -1;
    }
    return PerthTime.parseClockMinutes(journey.optString("preferredTrainTime", ""));
  }

  static int liveHorizonMinutes(JSONObject journey) {
    if (journey == null) {
      return 24 * 60;
    }
    int untilMinutes = PerthTime.parseClockMinutes(journey.optString("defaultUntil", ""));
    if (untilMinutes >= 0) {
      return untilMinutes;
    }
    return 24 * 60;
  }

  static boolean tripMatchesPreferredOrLater(
    JSONObject trip,
    int preferredMinutes,
    int horizonMinutes
  ) {
    String departureIso = tripDepartureIso(trip);
    int departureMinutes = PerthTime.minutesFromIso(departureIso);
    if (departureMinutes < 0) {
      return false;
    }
    if (departureMinutes < preferredMinutes) {
      return false;
    }
    // Same-day Active until only (preferred before until). Overnight windows skip horizon clip.
    if (
      horizonMinutes < 24 * 60 &&
      preferredMinutes < horizonMinutes &&
      departureMinutes >= horizonMinutes
    ) {
      return false;
    }
    return true;
  }

  static JSONArray collectUpcomingTrips(JSONObject payload) {
    JSONArray upcoming = payload.optJSONArray("upcoming");
    if (upcoming != null && upcoming.length() > 0) {
      return upcoming;
    }
    JSONArray built = new JSONArray();
    JSONObject next = payload.optJSONObject("next");
    if (next != null) {
      built.put(next);
    }
    JSONObject following = payload.optJSONObject("following");
    if (following != null) {
      built.put(following);
    }
    return built;
  }

  static JSONObject resolveFollowingTrip(JSONObject payload, JSONObject activeNext) {
    if (payload == null || activeNext == null) {
      return null;
    }

    String activeDeparture = tripDepartureIso(activeNext);
    JSONArray upcoming = payload.optJSONArray("upcoming");
    if (upcoming != null && upcoming.length() > 0) {
      JSONObject fromArray = followingFromUpcoming(upcoming, activeDeparture);
      if (fromArray != null) {
        return fromArray;
      }
    }

    JSONObject following = payload.optJSONObject("following");
    if (
      following != null &&
      !tripDepartureIso(following).equals(activeDeparture) &&
      !hasDepartureMinutePassed(following)
    ) {
      return following;
    }

    return null;
  }

  static JSONObject followingFromUpcoming(JSONArray upcoming, String activeDeparture) {
    int activeIndex = findTripIndex(upcoming, activeDeparture);
    if (activeIndex >= 0 && activeIndex + 1 < upcoming.length()) {
      JSONObject candidate = upcoming.optJSONObject(activeIndex + 1);
      if (candidate != null && !hasDepartureMinutePassed(candidate)) {
        return candidate;
      }
    }

    return firstUpcomingAfter(upcoming, activeDeparture);
  }

  static int findTripIndex(JSONArray upcoming, String activeDeparture) {
    for (int index = 0; index < upcoming.length(); index += 1) {
      JSONObject trip = upcoming.optJSONObject(index);
      if (trip != null && tripDepartureIso(trip).equals(activeDeparture)) {
        return index;
      }
    }
    return -1;
  }

  static JSONObject firstUpcomingAfter(JSONArray upcoming, String activeDeparture) {
    long activeMs = PerthTime.epochMillisFromIso(activeDeparture);
    if (activeMs <= 0L) {
      return null;
    }

    for (int index = 0; index < upcoming.length(); index += 1) {
      JSONObject trip = upcoming.optJSONObject(index);
      if (trip == null || hasDepartureMinutePassed(trip)) {
        continue;
      }
      String departureIso = tripDepartureIso(trip);
      if (departureIso.equals(activeDeparture)) {
        continue;
      }
      if (PerthTime.epochMillisFromIso(departureIso) > activeMs) {
        return trip;
      }
    }

    return null;
  }

  static String tripDepartureIso(JSONObject trip) {
    if (trip == null) {
      return "";
    }
    String departureIso = trip.optString("departure", "");
    if (!departureIso.isEmpty()) {
      return departureIso;
    }
    return trip.optString("arrival", "");
  }

  static boolean hasDepartureMinutePassed(JSONObject trip) {
    return hasDepartureMinutePassed(tripDepartureIso(trip));
  }

  static boolean hasDepartureMinutePassed(String departureIso) {
    long advanceAt = departureAdvanceAtMs(departureIso);
    return advanceAt > 0L && System.currentTimeMillis() >= advanceAt;
  }

  public static boolean needsLocalRepaint(JSONObject snapshot) {
    return CommuteScheduleSnapshot.needsLocalRepaint(snapshot);
  }

  public static boolean isLiveCommuteSnapshot(JSONObject snapshot) {
    return CommuteScheduleSnapshot.isLiveCommuteSnapshot(snapshot);
  }

  public static boolean shouldOpportunisticRefresh(JSONObject snapshot) {
    return CommuteScheduleSnapshot.shouldOpportunisticRefresh(snapshot);
  }

  public static boolean needsPreDeparturePrefetch(JSONObject snapshot) {
    return CommuteScheduleSnapshot.needsPreDeparturePrefetch(snapshot);
  }

  /** First instant after the scheduled departure minute — time to fetch the next train. */
  static long departureAdvanceAtMs(String departureIso) {
    if (departureIso == null || departureIso.isEmpty()) {
      return 0L;
    }
    long departureMs = PerthTime.epochMillisFromIso(departureIso);
    if (departureMs <= 0L) {
      return 0L;
    }
    return ((departureMs / 60_000L) + 1L) * 60_000L;
  }

  public static boolean needsNetworkRefresh(JSONObject snapshot) {
    return CommuteScheduleSnapshot.needsNetworkRefresh(snapshot);
  }

  static String getLeavePhase(int minutesUntilLeave, int minutesUntilDeparture) {
    if (minutesUntilDeparture <= 0) {
      return "missed";
    }
    if (minutesUntilLeave < 0) {
      return "late";
    }
    if (minutesUntilLeave <= 0) {
      return "now";
    }
    if (minutesUntilLeave <= 2) {
      return "urgent";
    }
    if (minutesUntilLeave <= 5) {
      return "soon";
    }
    return "calm";
  }

  static String degradedPrimary(JSONObject cached) {
    return CommuteSchedulePreview.degradedPrimary(cached);
  }
}

File: ios/Shared/CommuteSchedule.swift
import Foundation

enum CommuteSchedule {
    static let staleThresholdMs: Int64 = 120 * 60 * 1000

    struct Result {
        var settings: [String: Any]?
        var journey: [String: Any]?
        var payload: [String: Any]?
        var next: [String: Any]?
        var departMode = false
        var stale = false
        var empty = false
        var nearbyFallback = false
        var refreshedAtMs: Int64 = 0
        var journeyId = ""
        var route = ""
        var leaveByIso = ""
        var departureIso = ""
        var displayTime = ""
        var status = ""
        var leavePhase = ""
        var minutesUntilLeave = 0
    }

    static func load(allowStaleFallback: Bool) -> [String: Any] {
        var result = Result()
        result.refreshedAtMs = WidgetSettingsStore.readLastRefreshMs()

        guard let settingsJson = WidgetSettingsStore.readSettings(),
              let data = settingsJson.data(using: .utf8),
              let settings = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return emptyState()
        }

        result.settings = settings

        do {
            if let pinned = try WidgetPinResolver.loadBestPinnedResult(settings: settings) {
                var pinnedResult = pinned
                pinnedResult.refreshedAtMs = Int64(Date().timeIntervalSince1970 * 1000)
                WidgetSettingsStore.saveLastRefreshMs(pinnedResult.refreshedAtMs)
                pinnedResult.stale = false
                let snapshot = buildLiveSnapshot(pinnedResult)
                WidgetSettingsStore.saveSnapshot(snapshot)
                return snapshot
            }

            result.journey = JourneySelector.selectJourney(settings)

            if result.journey == nil {
                if JourneySelector.hasConfiguredJourneys(settings) {
                    return outsideHoursSnapshot(settings)
                }
                return emptyState()
            }

            guard let journey = result.journey else { return emptyState() }
        result.journeyId = journey["id"] as? String ?? ""
        result.route = WidgetDataService.formatRoute(journey)
        result.departMode = !(journey["useLeaveBefore"] as? Bool ?? true)
        let leaveBefore = result.departMode ? 0 : (journey["leaveBeforeMinutes"] as? Int ?? 10)

        do {
            result.payload = try NextTrainApiClient.fetchNextTrain(
                station: journey["station"] as? String ?? "",
                direction: journey["direction"] as? String ?? "",
                leaveBeforeMinutes: leaveBefore
            )
            result.refreshedAtMs = Int64(Date().timeIntervalSince1970 * 1000)
            WidgetSettingsStore.saveLastRefreshMs(result.refreshedAtMs)
            result.stale = false
            result.next = resolveActiveNextTrip(result.payload, journey: journey)
            fillTripFields(&result)
            let snapshot = buildLiveSnapshot(result)
            WidgetSettingsStore.saveSnapshot(snapshot)
            return snapshot
        } catch {
            if allowStaleFallback, let cached = WidgetSettingsStore.readSnapshot() {
                var snapshot = cached
                let age = Int64(Date().timeIntervalSince1970 * 1000) - result.refreshedAtMs
                if result.refreshedAtMs > 0 && age > staleThresholdMs {
                    snapshot["stale"] = true
                    snapshot["updatedLine"] = "Times may be out of date"
                }
                return repaintSnapshot(snapshot) ?? snapshot
            }
            return loadingState(journey)
        }
    }

    static func repaintSnapshot(_ cached: [String: Any]?) -> [String: Any]? {
        guard var snapshot = cached else { return nil }
        if snapshot["empty"] as? Bool == true ||
            snapshot["nearbyFallback"] as? Bool == true ||
            snapshot["outsideHoursIdle"] as? Bool == true {
            return snapshot
        }

        let departureIso = snapshot["departureIso"] as? String ?? ""
        guard !departureIso.isEmpty else { return snapshot }

        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        if hasDepartureMinutePassed(departureIso) {
            return snapshot
        }
        return repaintActiveSnapshot(&snapshot, nowMs: nowMs)
    }

    static func snapshotForDisplay() -> [String: Any] {
        if let cached = WidgetSettingsStore.readSnapshot(),
           let repainted = repaintSnapshot(cached) {
            return repainted
        }
        return load(allowStaleFallback: true)
    }

    private static func fillTripFields(_ result: inout Result) {
        guard let next = result.next else { return }
        result.leaveByIso = next["leaveBy"] as? String ?? ""
        result.departureIso = next["departure"] as? String ?? ""
        if result.departureIso.isEmpty {
            result.departureIso = next["arrival"] as? String ?? ""
        }
        result.displayTime = next["displayTime"] as? String ?? "—"
        result.status = next["status"] as? String ?? "On Time"
        result.leavePhase = next["leavePhase"] as? String ?? "calm"
        result.minutesUntilLeave = next["minutesUntilLeave"] as? Int ?? 0
    }

    private static func resolveActiveNextTrip(_ payload: [String: Any]?, journey: [String: Any]?) -> [String: Any]? {
        guard let payload else { return nil }
        let upcoming = payload["upcoming"] as? [[String: Any]] ?? []
        if !upcoming.isEmpty {
            for trip in upcoming where !hasDepartureMinutePassed(tripDepartureIso(trip)) {
                return trip
            }
            return nil
        }
        if let next = payload["next"] as? [String: Any], !hasDepartureMinutePassed(tripDepartureIso(next)) {
            return next
        }
        if let following = payload["following"] as? [String: Any],
           !hasDepartureMinutePassed(tripDepartureIso(following)) {
            return following
        }
        return nil
    }

    private static func buildLiveSnapshot(_ result: Result) -> [String: Any] {
        guard let journey = result.journey else { return emptyState() }

        if result.next == nil && result.payload == nil {
            return loadingState(journey)
        }

        if result.next == nil {
            return [
                "empty": false,
                "journeyId": result.journeyId,
                "route": result.route,
                "stationLabel": result.route,
                "label": "NEXT TRAIN",
                "primary": "No trains",
                "trainClock": "",
                "secondary": "",
                "updatedLine": PerthTime.formatUpdatedAgo(result.refreshedAtMs),
                "statusCrumb": "",
                "urgent": false,
                "late": false,
                "stale": result.stale,
            ]
        }

        guard let next = result.next else { return emptyState() }
        let minutesUntilDeparture = next["minutesUntilDeparture"] as? Int ?? 0
        let leavePhase = result.leavePhase
        let minutesUntilLeave = result.minutesUntilLeave
        let departMode = result.departMode
        let leaveArmed = departMode ? false : leaveByArmedForTrip(next, journey: journey)

        var leaveUrgent = ["now", "urgent", "soon"].contains(leavePhase)
        var secondary = ""
        if !departMode && leaveArmed {
            secondary = formatLeaveSecondary(leavePhase: leavePhase, minutesUntilLeave: minutesUntilLeave)
            if secondary == "Leave now" && minutesUntilLeave <= 0 { leaveUrgent = true }
        }

        return [
            "empty": false,
            "journeyId": result.journeyId,
            "route": result.route,
            "stationLabel": result.route,
            "journeyName": journey["name"] as? String ?? "Journey",
            "stale": result.stale,
            "label": liveWidgetLabel(journey: journey, trip: next),
            "primary": formatMinutesPrimary(minutesUntilDeparture),
            "trainClock": result.displayTime,
            "secondary": secondary,
            "urgent": leaveUrgent && !secondary.isEmpty,
            "late": false,
            "leaveByArmed": leaveArmed,
            "preferredHint": leaveArmed ? "" : preferredHintForJourney(journey),
            "statusCrumb": formatStatusCrumb(result.status),
            "departureIso": result.departureIso,
            "leaveByIso": result.leaveByIso,
            "departMode": departMode,
            "refreshedAtMs": result.refreshedAtMs,
            "status": result.status,
            "updatedLine": result.stale ? "Times may be out of date" : PerthTime.formatUpdatedAgo(result.refreshedAtMs),
            "leaveBeforeMinutes": journey["leaveBeforeMinutes"] as? Int ?? 10,
        ]
    }

    private static func repaintActiveSnapshot(_ snapshot: inout [String: Any], nowMs: Int64) -> [String: Any] {
        let departureIso = snapshot["departureIso"] as? String ?? ""
        let leaveByIso = snapshot["leaveByIso"] as? String ?? ""
        let minutesUntilDeparture = PerthTime.minutesUntilWallClock(departureIso, nowMs: nowMs)
        let minutesUntilLeave = leaveByIso.isEmpty
            ? minutesUntilDeparture
            : PerthTime.minutesUntilWallClock(leaveByIso, nowMs: nowMs)
        let leavePhase = getLeavePhase(minutesUntilLeave: minutesUntilLeave, minutesUntilDeparture: minutesUntilDeparture)
        let departMode = snapshot["departMode"] as? Bool ?? false
        let leaveArmed = snapshot["leaveByArmed"] as? Bool ?? true

        snapshot["label"] = "NEXT TRAIN"
        snapshot["primary"] = formatMinutesPrimary(minutesUntilDeparture)

        if departMode || !leaveArmed {
            snapshot["secondary"] = ""
            snapshot["urgent"] = false
            snapshot["late"] = false
        } else {
            let secondary = formatLeaveSecondary(leavePhase: leavePhase, minutesUntilLeave: minutesUntilLeave)
            snapshot["secondary"] = secondary
            snapshot["urgent"] = (["now", "urgent", "soon"].contains(leavePhase) || secondary == "Leave now") && !secondary.isEmpty
            snapshot["late"] = false
        }

        if let refreshedAtMs = snapshot["refreshedAtMs"] as? Int64 {
            let stale = snapshot["stale"] as? Bool ?? false
            snapshot["updatedLine"] = stale ? "Times may be out of date" : PerthTime.formatUpdatedAgo(refreshedAtMs)
        }
        return snapshot
    }

    private static func outsideHoursSnapshot(_ settings: [String: Any]) -> [String: Any] {
        let nowMinutes = PerthTime.minutesSinceMidnight()
        let day = PerthTime.dayOfWeekIso()
        guard let preview = NextCommutePreview.findNext(settings: settings, nowMinutes: nowMinutes, dayOfWeekIso: day) else {
            return nearbyFallbackState()
        }
        let route = WidgetDataService.formatRoute(preview.journey)
        return [
            "empty": false,
            "outsideHoursIdle": true,
            "openNearbyOnTap": true,
            "nearbyFallback": false,
            "journeyId": preview.journey["id"] as? String ?? "nearby",
            "route": route,
            "stationLabel": route,
            "journeyName": preview.journey["name"] as? String ?? "",
            "label": NextCommutePreview.idleWidgetLabel(preview.journey),
            "primary": preview.preferredOrFromClock,
            "trainClock": NextCommutePreview.formatDayWord(preview),
            "secondary": "",
            "preferredHint": "",
            "leaveByArmed": false,
            "updatedLine": "",
            "statusCrumb": "",
            "departureIso": "",
            "leaveByIso": "",
            "stale": false,
            "urgent": false,
            "late": false,
        ]
    }

    private static func nearbyFallbackState() -> [String: Any] {
        [
            "nearbyFallback": true,
            "empty": false,
            "journeyId": "nearby",
            "label": "NEAR ME",
            "primary": "Near me",
            "trainClock": "See trains near you",
            "secondary": "",
            "updatedLine": "",
            "stale": false,
            "urgent": false,
            "late": false,
        ]
    }

    private static func emptyState() -> [String: Any] {
        [
            "empty": true,
            "journeyId": "new",
            "label": "NEXT TRAIN",
            "primary": "Add a journey",
            "trainClock": "Tap to set up",
            "secondary": "",
            "updatedLine": "",
            "stale": false,
            "urgent": false,
            "late": false,
        ]
    }

    private static func loadingState(_ journey: [String: Any]) -> [String: Any] {
        [
            "empty": false,
            "journeyId": journey["id"] as? String ?? "",
            "route": WidgetDataService.formatRoute(journey),
            "stationLabel": WidgetDataService.formatRoute(journey),
            "label": "NEXT TRAIN",
            "primary": "…",
            "trainClock": "",
            "secondary": "",
            "updatedLine": "",
            "stale": false,
            "urgent": false,
            "late": false,
        ]
    }

    private static func tripDepartureIso(_ trip: [String: Any]) -> String {
        let departure = trip["departure"] as? String ?? ""
        if !departure.isEmpty { return departure }
        return trip["arrival"] as? String ?? ""
    }

    private static func hasDepartureMinutePassed(_ departureIso: String) -> Bool {
        let departureMs = PerthTime.epochMillisFromIso(departureIso)
        guard departureMs > 0 else { return false }
        return Int64(Date().timeIntervalSince1970 * 1000) >= departureMs
    }

    private static func hasDepartureMinutePassed(_ trip: [String: Any]) -> Bool {
        hasDepartureMinutePassed(tripDepartureIso(trip))
    }

    private static func leaveByArmedForTrip(_ trip: [String: Any], journey: [String: Any]) -> Bool {
        if JourneyPinHelper.isOverrideActiveToday(journey) { return true }
        let preferredMinutes = PerthTime.parseClockMinutes(journey["preferredTrainTime"] as? String ?? "")
        if preferredMinutes < 0 { return true }
        let departureMinutes = PerthTime.minutesFromIso(tripDepartureIso(trip))
        return departureMinutes >= preferredMinutes
    }

    private static func liveWidgetLabel(journey: [String: Any]?, trip: [String: Any]?) -> String {
        guard trip != nil else { return "NEXT TRAIN" }
        if let journey, (journey["id"] as? String) == NearbyPinHelper.journeyId {
            return "Pinned Train"
        }
        if JourneyPinHelper.isOverrideActiveToday(journey) {
            return "Target Train"
        }
        let preferredMinutes = PerthTime.parseClockMinutes(journey?["preferredTrainTime"] as? String ?? "")
        if preferredMinutes < 0 { return "NEXT TRAIN" }
        let departureMinutes = PerthTime.minutesFromIso(tripDepartureIso(trip!))
        return departureMinutes >= preferredMinutes ? "Target Train" : "NEXT TRAIN"
    }

    private static func preferredHintForJourney(_ journey: [String: Any]) -> String {
        let preferredMinutes = PerthTime.parseClockMinutes(journey["preferredTrainTime"] as? String ?? "")
        guard preferredMinutes >= 0 else { return "" }
        return "Target \(NextCommutePreview.formatClock(preferredMinutes))"
    }

    private static func getLeavePhase(minutesUntilLeave: Int, minutesUntilDeparture: Int) -> String {
        if minutesUntilLeave < 0 { return "late" }
        if minutesUntilLeave <= 0 { return "now" }
        if minutesUntilLeave <= 2 { return "urgent" }
        if minutesUntilLeave <= 5 { return "soon" }
        return "calm"
    }

    private static func formatMinutesPrimary(_ minutes: Int) -> String {
        if minutes <= 0 { return "NOW" }
        if minutes == 1 { return "1 min" }
        return "\(minutes) min"
    }

    private static func formatLeaveSecondary(leavePhase: String, minutesUntilLeave: Int) -> String {
        if leavePhase == "late" || leavePhase == "missed" {
            let minutesLate = minutesUntilLeave < 0 ? abs(minutesUntilLeave) : 0
            return minutesLate <= 1 ? "Leave now" : ""
        }
        if leavePhase == "now" || minutesUntilLeave <= 0 { return "Leave now" }
        if minutesUntilLeave == 1 { return "Leave in 1 min" }
        return "Leave in \(minutesUntilLeave) min"
    }

    private static func formatStatusCrumb(_ status: String) -> String {
        if status.isEmpty || status.caseInsensitiveCompare("On Time") == .orderedSame { return "" }
        return status
    }
}
